fil_fungible_token/src/token/state.rs

```diff
@@ -41,6 +44,32 @@ pub enum StateError {
     NegativeTotalSupply { supply: TokenAmount, delta: TokenAmount },
 }
 
+#[derive(Error, Debug)]
+pub enum StateInvariantError {
+    #[error("total supply was negative: {0}")]
+    SupplyNegative(TokenAmount),
+    #[error("the account for {account:?} had a negative balance of {balance:?}")]
+    BalanceNegative { account: ActorID, balance: TokenAmount },
+    #[error("the total supply {supply:?} does not match the sum of all balances {balance_sum:?}")]
+    BalanceSupplyMismatch { supply: TokenAmount, balance_sum: TokenAmount },
+    #[error(
+        "a negative allowance of {allowance:?} was specified between {owner:?} and {operator:?}"
+    )]
+    NegativeAllowance { owner: ActorID, operator: ActorID, allowance: TokenAmount },
+    #[error("stored a zero balance which should have been removed for {0}")]
+    ExplicitZeroBalance(ActorID),
+    #[error(
+        "stored a zero allowance which should have been removed between {owner:?} and {operator:?}"
+    )]
+    ExplicitZeroAllowance { owner: ActorID, operator: ActorID },
+    #[error("stored an allowance map for {0} though they have specified no allowances")]
+    ExplicitEmptyAllowance(ActorID),
+    #[error("stored an allowance for self {account:?} for {allowance:?}")]
+    ExplicitSelfAllowance { account: ActorID, allowance: TokenAmount },
+    #[error("underlying state error {0}")]
+    State(#[from] StateError),
+}
```

Comment:
Thanks. Since these are not really used in production, and any error is bad, this level of structure is probably overkill. But now that its here, great.